src/main.cpp

```diff
@@ -1472,14 +1483,16 @@ bool CheckInputs(const CTransaction& tx, CValidationState &state, const CCoinsVi
 
             // If prev is coinbase, check that it's matured
             if (coins->IsCoinBase()) {
-                if (nSpendHeight - coins->nHeight < COINBASE_MATURITY)
+                if (nSpendHeight - coins->nHeight < (protocol_cleanup ? 1 : COINBASE_MATURITY))
```

Comment:
Since `nSpendHeight` is always "latest height + 1", and since `protocol_cleanup` introduces a 1 height coinbase spend, you can just do

```suggestion
                if (!protocol_cleanup && nSpendHeight - coins->nHeight < COINBASE_MATURITY)
```

Reply:
I think I’m checking something different here. Keep in mind that the coin view is being updated during block validation, so the coinbase outputs winter the UTXO set and would be spendable by transactions in the same block.

Now to validate the miner would then have to leave any excess beyond the subsidy as fee later in the block, but not before doing some damage. With asset issuance they could send themselves an issuance token, which they use before returning it to the coinbase as fee.

Without asset issuance there aren’t any catastrophic failures I can think of, but it does remove some error margins... if there are any undiscovered overflow possibilities for example, they would become exploitable by anyone able to solo-mine a block.

So to protect against all that I think we need a maturity period of at least one block—the earliest you can spend a coinbase output is the block *after*.

Does that make sense? Or did I miss your point? I’m on the go on mobile right now and may have misunderstood :)

Reply:
I think you're missing my point. `nSpendHeight - coins->nHeight` will *never* be `< 1`, so you might as well skip the check if `protocol_cleanup` is set. Unless you intend to tune that `1`. Which sounds dangerous.

Reply:
`nSpendHeight - coins->nHeight` will be `< 1` for the spends of dependent transactions in the same block.

Reply:
There are no dependent transactions in the same block that are simultaneously coin base transactions.

Reply:
It is checking if the output being spent is from a coinbase.

Reply:
Wait, so this ends up checking the single case where the miner creates a coinbase output, and then immediately spends the coinbase transaction inside the block itself. I.e.

```
tx 0 (coinbase) -> x
tx 1 (same block) x -> y 
```
? I have no idea why you even care about that, as I believe coinbases can have multiple outputs as it is (so doing tx1 above is literally pointless, if maturity = 1).

That aside, I at least understand what the code is forbidding now, even if I don't really understand the "why". :)

Reply:
(Moving this response to the review thread, since for some reason Github put the email reply elsewhere in the UI.)

Tx 0 (coinbase) -> x (= issuance token for asset A)
Tx 1 (same block) x -> y (and issue some A)
Tx 2 (same block) y -> (return as fee to the coinbase)

In this example the miner creates an output in his coinbase of an asset he has no right to own, which is an issuance token for another asset. He then uses it to issue the asset, then returns the issuance token to the coinbase as fee. From the block’s perspective the coinbase didn’t generate anything which wasn’t given to it as fee, but that extra issuance token does exist for the duration of the block.

Alternatively, the miner issues a bajillion coins to himself, gets to use them for the duration of the block, before returning them all back to himself as fee. By itself this isn’t useful, but if there was another undiscovered overflow bug it could be catastrophic.

In short, the issue is that the miner can print money that shouldn’t exist, use it for the duration of the block only, and then return it in a time traveling payment back to himself at the end of the block, as fee. This should probably be outlawed just for being weird and unexpected, but could actually cause problems too with new features (asset issuance) or undiscovered bugs (overflow).

Requiring 1 block maturity for coinbases prevents this.